scripts/build/build/factory.py

```diff
@@ -0,0 +1,127 @@
+import os
+
+from typing import Set
+
+from builders.builder import Builder
+from builders.linux import LinuxBuilder
+from builders.qpg import QpgBuilder
+from builders.esp32 import Esp32Builder, Esp32Board
+from builders.efr32 import Efr32Builder, Efr32App, Efr32Board
+
+from .targets import Application, Board, Platform
+
+
+class Matcher():
+  """Figures out if a proper builder can be created for a platform/board/app combination."""
+
+  def __init__(self, builder_class):
+    self.builder_class = builder_class
+    self.app_arguments = {}
+    self.board_arguments = {}
+
+  def AcceptApplication(self, __app_key: Application, **kargs):
+    self.app_arguments[__app_key] = kargs
+
+  def AcceptBoard(self, __board_key: Board, **kargs):
+    self.board_arguments[__board_key] = kargs
+
+  def Create(self, __board_key: Board, __app_key: Application, repo_path: str,
+             **kargs):
+    """Creates a new builder for the given board/app. """
+    if not __board_key in self.board_arguments:
+      return None
+
+    if not __app_key in self.app_arguments:
+      return None
+
+    kargs.update(self.board_arguments[__board_key])
+    kargs.update(self.app_arguments[__app_key])
+
+    return self.builder_class(repo_path, **kargs)
+
+
+# Builds a list of acceptable application/board combination for every platform
+_MATCHERS = {
+    Platform.LINUX: Matcher(LinuxBuilder),
+    Platform.ESP32: Matcher(Esp32Builder),
+    Platform.QPG: Matcher(QpgBuilder),
+    Platform.EFR32: Matcher(Efr32Builder),
+}
+
+# Matrix of what can be compiled using and what build options are required
+# by such compilation
+_MATCHERS[Platform.LINUX].AcceptApplication(Application.ALL_CLUSTERS)
+_MATCHERS[Platform.LINUX].AcceptBoard(Board.NATIVE)
+
+_MATCHERS[Platform.ESP32].AcceptApplication(Application.ALL_CLUSTERS)
```

Comment:
ESP32 can build lock-app, no?

I wish we just auto-determined this from the directory structure....  followup?

Reply:
Agree, will need to add more as a followup (want chip-tool, shell and other example apps).

Initial commit intent is showing that the structure can support diverse targers so I picked a set of example applications and platforms that seemed diverse enough.

Reply:
I also added the lock app as a proof of concept.

This shows another strangeness: we allow all-clusters build for both m5stack and devkitc, but lock is only devkitc.
Updated code handles that too.

Reply:
Testing seems to work well:

```
vscode@andreilitvin:/workspaces/connectedhomeip$ ./scripts/build/chipbuild.py --app lock gen >/dev/null 2>&1;
ls -l out
vscode@andreilitvin:/workspaces/connectedhomeip$ ls -l out
total 12
drwx------ 6 vscode vscode 4096 Jul  9 14:54 efr32-brd4161a-lock
drwxrwxr-x 7 vscode vscode 4096 Jul  9 14:54 esp32-devkitc-lock
drwx------ 4 vscode vscode 4096 Jul  9 14:54 qpg-qpg6100-lock
vscode@andreilitvin:/workspaces/connectedhomeip$ 
```